Add React error overlay and HMR source maps

devongovett opened the pull request:
Closes #7604. Closes #643. Closes #5316.

<img width="1198" alt="image" src="https://user-images.githubusercontent.com/19409/166131688-ddc05e6f-7491-4be7-9ea7-1524b9c26a8d.png">

This adds an overlay that appears when there is a runtime error in a React app, using the [react-error-overlay](https://github.com/facebook/create-react-app/tree/a422bf227cf5294a34d68696664e9568a152fd8f/packages/react-error-overlay) package from Create React App. It is implemented as part of `@parcel/runtime-react-refresh` since it is React-specific, and that seemed like the easiest place to put it. It will appear when a runtime error occurs, and is automatically dismissed when an HMR update comes in.

This required support for source maps in HMR, which has been a longstanding issue due to our use of `new Function`, which has poor support for source maps. Now, we use `eval` instead, which has better support in Chrome and Firefox. However, Safari still doesn't support `sourceMappingURL` or `sourceURL` in `eval` error stack traces, so we detect this case and fall back to loading the asset from a new dev server endpoint that serves asset contents by id.

As a side benefit, HMR should now fall back to loading URLs for changed assets via HTTP if `eval` is not allowed, e.g. with a strict CSP. This may help web extension v3 support, though I have not tested this. (cc. @101arrowz)

The error overlay also supports clicking the code frame to open it directly in your editor, which I also added to the standard parcel build error overlay and the error 500 page. This is implemented through a second new dev server endpoint which auto-detects which editor you're using, and opens it to the line and column of the error message.

Finally, the error 500 page has been improved to automatically reload when an HMR update comes in so you don't need to manually do that.

devongovett committed:
Fix source maps in HMR

devongovett committed:
Add react-error-overlay

devongovett committed:
Add ability to launch editor from browser

devongovett committed:
Automatically reload error page via HMR

devongovett committed:
Update yarn.lock

devongovett committed:
lint

devongovett commented on packages/core/utils/src/prettyDiagnostic.js:
```diff
@@ -10,10 +10,18 @@ import nullthrows from 'nullthrows';
+  /** A formatted string containing all code frames, including their file locations. */
   codeframe: string,
+  /** A list of code frames with highlighted code and file locations separately. */
+  frames: Array<FormattedCodeFrame>,
```
A bit unfortunate, but couldn't think of a better way...

101arrowz replied:
You could make `codeframe` a getter that processes `frames` in the default manner so there's less duplication. Beyond disgusting regex I don't think there's any other way to support rewriting paths than saving the `frames` array.

devongovett commented on packages/reporters/dev-server/src/HMRServer.js:
```diff
@@ -185,3 +201,34 @@ function getSpecifier(dep: Dependency): string {
+  if (sourcemap) {
+    let sourcemapStringified = await sourcemap.stringify({
+      format: 'inline',
+      sourceRoot: '/__parcel_source_root/',
```
Somewhat related, but this source root seems to be displayed by react-error-overlay rather than either the original file path, or nothing. Not sure if there is a way to improve that?

101arrowz commented:
Yes, this should help manifest V3 hot reload. As long as the HMR server is on some kind of localhost domain, dynamic script loading should work fine via a similar CSP patching technique to MV2. I'll look into it more after this is merged.

101arrowz commented on packages/runtimes/hmr/src/loaders/hmr-runtime.js:
```diff
@@ -92,8 +92,18 @@ if ((!parent || !parent.isParcelRequire) && typeof WebSocket !== 'undefined') {
+
   // $FlowFixMe
-  ws.onmessage = function (event /*: {data: string, ...} */) {
+  ws.onmessage = async function (event /*: {data: string, ...} */) {
```
Does using `async/await` have any consequences for developing in old browsers? AFAICT most of our other runtimes are ES5 syntactically.

devongovett replied:
It should get transpiled I believe.

101arrowz commented on packages/runtimes/hmr/src/loaders/hmr-runtime.js:
```diff
@@ -120,9 +130,15 @@ if ((!parent || !parent.isParcelRequire) && typeof WebSocket !== 'undefined') {
+          typeof window !== 'undefined' &&
+          typeof CustomEvent !== 'undefined'
+        ) {
+          window.dispatchEvent(new CustomEvent('parcelhmraccept'));
```
Minor thing but you might instead use `parcel-hmr-accept` in styling convention with DOM events. Also since you're already using global state you could also a well-known symbol to `global` containing a `registerAcceptor` function or something to avoid polluting user code.

devongovett replied:
hmm I think most DOM events actually don't use dashes? And the CustomEvent example on MDN also doesn't... https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent/CustomEvent. Anyway, doesn't really matter. This isn't meant to be public API.

parcel-benchmark commented:
## Benchmark Results
<details><summary>Kitchen Sink 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 8.79s | +58.00ms |
| Cached | 445.00ms | +14.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/logo.c5bb83f1.png | 246.00b | +0.00b | 3.84s | -581.00ms 🚀 |

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 1.50m | -666.00ms |
| Cached | 2.41s | -92.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.0605f25b.js | 3.32mb | +144.00b ⚠️ | 29.70s | -349.00ms |
| dist/16.069344b7.js | 905.00b | +0.00b | 33.73s | -3.84s 🚀 |
| dist/simpleHasher.46d6f2e5.js | 742.00b | +0.00b | 37.29s | -31.24s 🚀 |
| dist/index.html | 240.00b | +0.00b | 29.71s | -3.31s 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/esm.6190c9ac.js | 61.54kb | +0.00b | 1.13m | +30.73s ⚠️ |
| dist/workerHasher.e50d242f.js | 1.72kb | +0.00b | 1.13m | +30.73s ⚠️ |
| dist/16.1969624f.js | 1.08kb | +0.00b | 36.72s | +3.07s ⚠️ |
| dist/16.069344b7.js | 905.00b | +0.00b | 36.72s | +3.07s ⚠️ |
| dist/simpleHasher.46d6f2e5.js | 742.00b | +0.00b | 1.13m | +30.73s ⚠️ |
| dist/index.html | 240.00b | +0.00b | 1.13m | +38.42s ⚠️ |

</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 6.04s | -26.00ms |
| Cached | 297.00ms | +25.00ms ⚠️ |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/330481156792255059)**

devongovett committed:
Remove .babelrc from hmr-runtime

It relies on regeneratorRuntime which isn't always defined, and SWC will compile the runtime anyway

101arrowz commented:
The logic here for the dev server is a bit problematic for web extensions since it depends on `parcel serve` rather than just `parcel watch` in order to serve the new chunks. To fix this we'd need to refactor a lot of serving logic. Additionally, some of the hardcoded `/__parcel_source_root` won't work because it will try to access the web extension context, we need to change it into absolute paths including hostname if `serveOptions.host != null`.

I'll try to do this at some point.
